js/adapt-stateful-session.js

```diff
@@ -1,261 +1,266 @@
-      // The config allows the user to override the completion state.
-      switch (completionData.status) {
-        case COMPLETION_STATE.COMPLETED:
-        case COMPLETION_STATE.PASSED: {
-          if (!config || !config._reporting || !config._reporting._onTrackingCriteriaMet) {
-            Adapt.log.warn(`No value defined for '_onTrackingCriteriaMet', so defaulting to '${completionStatus}'`);
-          } else {
-            completionStatus = config._reporting._onTrackingCriteriaMet;
-          }
+  onLanguageChanged() {
+    // when the user switches language, we need to:
+    // - reattach the event listeners as the language change triggers a reload of
+    //   the json, which will create brand new collections
+    // - get and save a fresh copy of the session state. as the json has been reloaded,
+    //   the blocks completion data will be reset (the user is warned that this will
+    //   happen by the language picker extension)
+    // - check to see if the config requires that the lesson_status be reset to
+    //   'incomplete'
+    const config = Adapt.spoor.config;
+    this.removeEventListeners();
+    this.setupEventListeners();
+    this.saveSessionState();
+    if (config?._reporting?._resetStatusOnLanguageChange === false) return;
+    Adapt.offlineStorage.set('status', 'incomplete');
+  }
 
-          break;
-        }
-        case COMPLETION_STATE.FAILED: {
-          if (!config || !config._reporting || !config._reporting._onAssessmentFailure) {
-            Adapt.log.warn(`No value defined for '_onAssessmentFailure', so defaulting to '${completionStatus}'`);
-          } else {
-            completionStatus = config._reporting._onAssessmentFailure;
-          }
+  onVisibilityChange() {
+    if (document.visibilityState === 'hidden') this.scorm.commit();
+  }
+
+  onQuestionRecordInteraction(questionView) {
+    if (!this._shouldRecordInteractions) return;
+    const responseType = questionView.getResponseType();
+    // If responseType doesn't contain any data, assume that the question
+    // component hasn't been set up for cmi.interaction tracking
+    if (_.isEmpty(responseType)) return;
+    const id = questionView.model.get('_id');
+    const response = questionView.getResponse();
+    const result = questionView.isCorrect();
+    const latency = questionView.getLatency();
+    Adapt.offlineStorage.set('interaction', id, response, result, latency, responseType);
+  }
+
+  onAssessmentComplete(stateModel) {
+    const config = Adapt.spoor.config;
+    Adapt.course.set('_isAssessmentPassed', stateModel.isPass);
+    this.saveSessionState();
+    const shouldSubmitScore = (config?._tracking?._shouldSubmitScore);
+    if (!shouldSubmitScore) return;
+    const scoreArgs = stateModel.isPercentageBased ?
+      [ stateModel.scoreAsPercent, 0, 100 ] :
+      [ stateModel.score, 0, stateModel.maxScore ];
+    Adapt.offlineStorage.set('score', ...scoreArgs);
+  }
+
```

Comment:
This stuff has just been moved out.
https://github.com/adaptlearning/adapt-contrib-spoor/pull/222